### PR TITLE
support truncate partitioned table and drop partitioned table

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)
support truncate partitioned table
* the partition table deletes the partition data, but retains the `partitionInfo`.
* each partition reassigns the `partitionID`.

support drop partitioned table
* delete partition data is added.
```sql
tidb> set @@session.tidb_enable_table_partition=1;
Query OK, 0 rows affected (0.00 sec)

tidb> create table tr (id int, name varchar(50), purchased date)
    ->     partition by range( year(purchased) ) (
    ->     partition p0 values less than (1990),
    ->     partition p1 values less than (1995),
    ->     partition p2 values less than (2000),
    ->     partition p3 values less than (2005),
    ->     partition p4 values less than (2010),
    ->     partition p5 values less than (2015)
    ->  );
Query OK, 0 rows affected (0.01 sec)

tidb> insert into tr values
    ->    (1, 'desk organiser', '2003-10-15'),
    ->    (2, 'alarm clock', '1997-11-05'),
    ->    (3, 'chair', '2009-03-10'),
    ->    (4, 'bookcase', '1989-01-10'),
    ->    (5, 'exercise bike', '2014-05-09'),
    ->    (6, 'sofa', '1987-06-05'),
    ->    (7, 'espresso maker', '2011-11-22'),
    ->    (8, 'aquarium', '1992-08-04'),
    ->    (9, 'study desk', '2006-09-16'),
    ->    (10, 'lava lamp', '1998-12-25');
Query OK, 10 rows affected (0.12 sec)
```
master branch
```sql
tidb> truncate table tr;
Query OK, 0 rows affected (0.01 sec)

tidb> select * from tr;
+------+----------------+------------+
| id   | name           | purchased  |
+------+----------------+------------+
|    8 | aquarium       | 1992-08-04 |
|    3 | chair          | 2009-03-10 |
|    9 | study desk     | 2006-09-16 |
|    2 | alarm clock    | 1997-11-05 |
|   10 | lava lamp      | 1998-12-25 |
|    1 | desk organiser | 2003-10-15 |
|    5 | exercise bike  | 2014-05-09 |
|    7 | espresso maker | 2011-11-22 |
|    4 | bookcase       | 1989-01-10 |
|    6 | sofa           | 1987-06-05 |
+------+----------------+------------+
10 rows in set (0.00 sec)
```

truncate-table-partition branch
```sql
tidb> truncate table tr;
Query OK, 0 rows affected (0.01 sec)

tidb> select * from tr;
Empty set (0.00 sec)
```

## What is the type of the changes? (mandatory)

New feature
- New feature (non-breaking change which adds functionality)
- Improvement (non-breaking change which is an improvement to an existing feature)
- Bug fix (non-breaking change which fixes an issue)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this PR been tested? (mandatory)
unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Does this PR need to be added to the release notes? (mandatory)
Yes

### CODE REVIEW

**Comment by ciscoxll** [comment]
@zimulala @winkyao @tiancaiamao PTAL.

File: ddl/delete_range.go [hunk 1]
```diff
@@ -238,6 +238,20 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
 		}
 	case model.ActionDropTable, model.ActionTruncateTable:
 		tableID := job.TableID
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/delete_range.go]: @tiancaiamao Yes, ignore the old code.

File: ddl/table.go [hunk 2]
```diff
@@ -208,7 +206,33 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
 	}
+	oldPid := getPartitionIDs(tblInfo)
 	tblInfo.ID = newTableID
+	var newPi *model.PartitionInfo
+	if tblInfo.GetPartitionInfo() != nil {
+		newPi = &model.PartitionInfo{
+			Type:    tblInfo.Partition.Type,
+			Expr:    tblInfo.Partition.Expr,
+			Columns: tblInfo.Partition.Columns,
+			Enable:  true,
```
Comment (tiancaiamao) [inline ddl/table.go]: Enable should be the same with old partition Info.

File: ddl/table.go [hunk 3]
```diff
@@ -208,7 +206,33 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+				ID:       pid,
+				Name:     def.Name,
+				LessThan: def.LessThan,
+				Comment:  def.Comment,
+			}
+			newPi.Definitions = append(newPi.Definitions, piDef)
+		}
+	}
+	if newPi != nil {
+		tblInfo.Partition = newPi
```
Comment (tiancaiamao) [inline ddl/table.go]: This line can be move to 231.

File: ddl/table.go [hunk 4]
```diff
@@ -208,7 +206,33 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+	oldPid := getPartitionIDs(tblInfo)
 	tblInfo.ID = newTableID
```
Comment (tiancaiamao) [inline ddl/table.go]: I suggest not to touch the old table Info here, before the new table info is ready.
If you change something and meet errors later, what will happen?

File: ddl/delete_range.go [hunk 5]
```diff
@@ -238,6 +238,20 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
...
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&partitionIDs); err != nil {
+			return errors.Trace(err)
```
Comment (tiancaiamao) [inline ddl/delete_range.go]: Maybe we can try to decode it as []byte again here, to adapt to the old TiDB binary? @zimulala

File: ddl/table.go (hunk 4 above)
Comment (ciscoxll) [reply ddl/table.go]: @tiancaiamao This just saves the partitionID of the old table, no old table modified.

File: ddl/delete_range.go (hunk 5 above)
Comment (zimulala) [reply ddl/delete_range.go]: @tiancaiamao 
I think we can use the method of appending `partitionIDs` to Args directly.

File: ddl/table.go (hunk 4 above)
Comment (tiancaiamao) [reply ddl/table.go]: Error handling is not that simple, please take a close look.

```
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)	
```

File: ddl/delete_range.go (hunk 5 above)
Comment (ciscoxll) [reply ddl/delete_range.go]: @tiancaiamao @zimulala Done.

File: ddl/partition.go [hunk 6]
```diff
@@ -216,3 +217,24 @@ func onDropTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+		return []int64{}
+	}
+	partitionIDs := make([]int64, 0, len(table.Partition.Definitions))
+	for _, def := range table.Partition.Definitions {
+		partitionIDs = append(partitionIDs, def.ID)
+	}
+	return partitionIDs
+}
+
+func genGlobalID(d *ddlCtx) (int64, error) {
```
Comment (shenli) [inline ddl/partition.go]: There is a function [here](https://github.com/pingcap/tidb/blob/master/ddl/ddl.go#L403). Can you merge them together?

File: ddl/table.go [hunk 7]
```diff
@@ -128,8 +128,7 @@ func onDropTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		}
 		// Finish this job.
 		job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
-		startKey := tablecodec.EncodeTablePrefix(tableID)
-		job.Args = append(job.Args, startKey)
+		job.Args = append(job.Args, getPartitionIDs(tblInfo))
```
Comment (shenli) [inline ddl/table.go]: Is this a compatible change?

File: ddl/partition.go (hunk 6 above)
Comment (ciscoxll) [reply ddl/partition.go]: @zimulala @tiancaiamao Do you think you can merge?

Comment (tiancaiamao) [reply ddl/partition.go]: If you use `getGlobalID(store kv.Storage)`, they can actually be one.
You can still keep `func (d *ddl) genGlobalID() (int64, error)`, let it call your if there are still many places use it and you don't want irrelevant changes in this PR.

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/partition.go (hunk 6 above)
Comment (shenli) [reply ddl/partition.go]: You can solve this in another PR.

File: ddl/delete_range.go [hunk 8]
```diff
@@ -238,7 +238,24 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
 		}
 	case model.ActionDropTable, model.ActionTruncateTable:
 		tableID := job.TableID
-		startKey := tablecodec.EncodeTablePrefix(tableID)
+		// The startKey here is for compatibility with previous versions.
+		var startKey kv.Key
```
Comment (shenli) [inline ddl/delete_range.go]: Why not treat the `endKey` in the same way?

File: ddl/table.go [hunk 9]
```diff
@@ -129,7 +129,7 @@ func onDropTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		// Finish this job.
 		job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
 		startKey := tablecodec.EncodeTablePrefix(tableID)
-		job.Args = append(job.Args, startKey)
+		job.Args = append(job.Args, startKey, getPartitionIDs(tblInfo))
```
Comment (shenli) [inline ddl/table.go]: Is this compatible with old version tidb-server?

File: ddl/table.go [hunk 10]
```diff
@@ -208,6 +208,31 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+	oldPartitionIDs := getPartitionIDs(tblInfo)
+	var newPi *model.PartitionInfo
```
Comment (shenli) [inline ddl/table.go]: Move this line into the if statement.

File: ddl/table.go (hunk 9 above)
Comment (ciscoxll) [reply ddl/table.go]: Old versions of tidb-server are compatible.

File: ddl/delete_range.go (hunk 8 above)
Comment (ciscoxll) [reply ddl/delete_range.go]: @shenli Because the old version did not handle ` endKey `.

File: ddl/table.go [hunk 11]
```diff
@@ -208,6 +208,32 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+	if tblInfo.GetPartitionInfo() != nil {
+		oldPartitionIDs = getPartitionIDs(tblInfo)
+		newPi := &model.PartitionInfo{
+			Type:    tblInfo.Partition.Type,
+			Expr:    tblInfo.Partition.Expr,
+			Columns: tblInfo.Partition.Columns,
+			Enable:  tblInfo.Partition.Enable,
+		}
+		for _, def := range tblInfo.Partition.Definitions {
+			pid, err := genGlobalID(d)
```
Comment (zimulala) [inline ddl/table.go]: Why don't we call `GenGlobalID()` directly?

File: ddl/delete_range.go [hunk 12]
```diff
@@ -238,7 +238,24 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
...
+		var partitionIDs []int64
+
```
Comment (zimulala) [inline ddl/delete_range.go]: Remove this blank line.

File: ddl/table.go [hunk 13]
```diff
@@ -208,6 +208,32 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return ver, errors.Trace(err)
+			}
+			piDef := model.PartitionDefinition{
```
Comment (zimulala) [inline ddl/table.go]: Cloud we just reset `pid`？ Then we needn't create `PartitionInfo` and `PartitionDefinition`.

File: ddl/db_test.go [hunk 14]
```diff
@@ -2629,3 +2629,94 @@ func (s *testDBSuite) TestAlterTableDropPartition(c *C) {
...
+	(6, 'sofa', '1987-06-05'),
+	(7, 'espresso maker', '2011-11-22'),
+	(8, 'aquarium', '1992-08-04'),
+	(9, 'study desk', '2006-09-16'),
+	(10, 'lava lamp', '1998-12-25');`)
+	result = s.tk.MustQuery("select * from t3;")
+	c.Assert(len(result.Rows()), Equals, 10)
+	s.tk.MustExec("truncate table t3;")
+	result = s.tk.MustQuery("select * from t3;")
+	c.Assert(len(result.Rows()), Equals, 0)
```
Comment (zimulala) [inline ddl/db_test.go]: Could we need to check if the operation of deleting ranges is complete, like `TestDropIndex`?

File: ddl/db_test.go [hunk 15]
```diff
@@ -2629,3 +2629,94 @@ func (s *testDBSuite) TestAlterTableDropPartition(c *C) {
...
+	(5, 'exercise bike', '2014-05-09'),
+	(6, 'sofa', '1987-06-05'),
+	(7, 'espresso maker', '2011-11-22'),
+	(8, 'aquarium', '1992-08-04'),
+	(9, 'study desk', '2006-09-16'),
+	(10, 'lava lamp', '1998-12-25');`)
+
+	result = s.tk.MustQuery("select * from t4;")
+	c.Assert(len(result.Rows()), Equals, 10)
+	s.tk.MustExec("drop table t4;")
```
Comment (zimulala) [inline ddl/db_test.go]: Ditto.

File: ddl/db_test.go [hunk 16]
```diff
@@ -2629,3 +2629,94 @@ func (s *testDBSuite) TestAlterTableDropPartition(c *C) {
...
+	(2, 'alarm clock', '1997-11-05'),
+	(3, 'chair', '2009-03-10'),
+	(4, 'bookcase', '1989-01-10'),
+	(5, 'exercise bike', '2014-05-09'),
+	(6, 'sofa', '1987-06-05'),
+	(7, 'espresso maker', '2011-11-22'),
+	(8, 'aquarium', '1992-08-04'),
+	(9, 'study desk', '2006-09-16'),
+	(10, 'lava lamp', '1998-12-25');`)
+
```
Comment (zimulala) [inline ddl/db_test.go]: Remove this blank line.

**Comment by winkyao** [comment]
@ciscoxll Plz fix ci.

File: ddl/delete_range.go (hunk 8 above)
Comment (zhexuany) [reply ddl/delete_range.go]: You probably need to put "Because the old version did not handle endKey." to comment too.

File: ddl/partition.go [hunk 17]
```diff
@@ -223,3 +223,14 @@ func checkAddPartitionTooManyPartitions(piDefs int) error {
 	}
 	return nil
 }
+
+func getPartitionIDs(table *model.TableInfo) []int64 {
```
Comment (zhexuany) [inline ddl/partition.go]: @tiancaiamao merged a PR with an exported method which allows you can get pid form that method.  By any chance, you could use that method.

File: ddl/table.go [hunk 18]
```diff
@@ -213,6 +213,25 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+
+	var oldPartitionIDs []int64
+	if tblInfo.GetPartitionInfo() != nil {
+		oldPartitionIDs = getPartitionIDs(tblInfo)
+		for _, def := range tblInfo.Partition.Definitions {
+			var pid int64
+			err := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply ddl/table.go]:
```
genID := func (txn kv,Transaction) error {
                var err error
				pid, err = meta.NewMeta(txn).GenGlobalID()
				return errors.Trace(err)
}
err := kv.RunInNewTxn(d.store, true, genID)
```

[the remark this answers is not in the document]
Comment (zhexuany) [reply ddl/table.go]: that way make code more self-explanatory.

File: ddl/table.go [hunk 19]
```diff
@@ -213,6 +213,26 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+	if tblInfo.GetPartitionInfo() != nil {
+		oldPartitionIDs = getPartitionIDs(tblInfo)
+		for _, def := range tblInfo.Partition.Definitions {
+			var pid int64
+			genID := func(txn kv.Transaction) error {
+				var err error
+				pid, err = meta.NewMeta(txn).GenGlobalID()
+				return errors.Trace(err)
+			}
+			err = kv.RunInNewTxn(d.store, true, genID)
```
Comment (zhexuany) [inline ddl/table.go]: replace `err = kv.RunInNewTxn(d.store, true, genID` with `err = kv.RunInNewTxn(d.store, true /* retryable */ , genID`

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL

File: ddl/db_test.go [hunk 20]
```diff
@@ -2672,3 +2672,133 @@ func (s *testDBSuite) TestAddPartitionTooManyPartitions(c *C) {
...
+func (s *testDBSuite) TestTruncatePartitionAndDropTable(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test;")
+	// Test truncate common table.
+	s.tk.MustExec("drop table if exists t1;")
+	s.tk.MustExec("create table t1 (id int(11));")
+	for i := 0; i < 100; i++ {
+		s.mustExec(c, "insert into t1 values (?)", i)
+	}
+	result := s.tk.MustQuery("select * from t1;")
```
Comment (winkyao) [inline ddl/db_test.go]: why not use `select count(*)`

File: ddl/db_test.go [hunk 21]
```diff
@@ -2672,3 +2672,133 @@ func (s *testDBSuite) TestAddPartitionTooManyPartitions(c *C) {
...
+	c.Assert(len(result.Rows()), Equals, 100)
+	s.tk.MustExec("truncate table t1;")
+	result = s.tk.MustQuery("select * from t1")
```
Comment (winkyao) [inline ddl/db_test.go]: ditto

File: ddl/db_test.go [hunk 22]
```diff
@@ -2672,3 +2672,133 @@ func (s *testDBSuite) TestAddPartitionTooManyPartitions(c *C) {
...
+	(3, 'chair', '2009-03-10'),
+	(4, 'bookcase', '1989-01-10'),
+	(5, 'exercise bike', '2014-05-09'),
+	(6, 'sofa', '1987-06-05'),
+	(7, 'espresso maker', '2011-11-22'),
+	(8, 'aquarium', '1992-08-04'),
+	(9, 'study desk', '2006-09-16'),
+	(10, 'lava lamp', '1998-12-25');`)
+	result = s.tk.MustQuery("select * from t3;")
+	c.Assert(len(result.Rows()), Equals, 10)
```
Comment (winkyao) [inline ddl/db_test.go]: ditto

File: ddl/db_test.go [hunk 23]
```diff
@@ -2672,3 +2672,133 @@ func (s *testDBSuite) TestAddPartitionTooManyPartitions(c *C) {
...
+	(2, 'alarm clock', '1997-11-05'),
+	(3, 'chair', '2009-03-10'),
+	(4, 'bookcase', '1989-01-10'),
+	(5, 'exercise bike', '2014-05-09'),
+	(6, 'sofa', '1987-06-05'),
+	(7, 'espresso maker', '2011-11-22'),
+	(8, 'aquarium', '1992-08-04'),
+	(9, 'study desk', '2006-09-16'),
+	(10, 'lava lamp', '1998-12-25');`)
+	result = s.tk.MustQuery("select * from t4;")
```
Comment (winkyao) [inline ddl/db_test.go]: ditto

File: ddl/db_test.go [hunk 24]
```diff
@@ -2672,3 +2672,133 @@ func (s *testDBSuite) TestAddPartitionTooManyPartitions(c *C) {
 	);`
 	s.testErrorCode(c, sql3, tmysql.ErrTooManyPartitions)
 }
+
+func checkPartitionDelRangeDone(c *C, s *testDBSuite, partitionPrefix kv.Key) bool {
+	hasOldPartitionData := true
+	for i := 0; i < waitForCleanDataRound; i++ {
```
Comment (winkyao) [inline ddl/db_test.go]: How can we make sure that the GC is done after waitForCleanDataRound * waitForCleanDataInterval time?

Comment (winkyao) [reply ddl/db_test.go]: GC interval is `gcWorkerTickInterval  = time.Minute`.

Comment (ciscoxll) [reply ddl/db_test.go]: @winkyao 
1.  `waitForCleanDataRound` indicates how many times should we check data is cleaned or not.
2.  `waitForCleanDataInterval` is a min duration between 2 check for data clean.

File: ddl/delete_range.go [hunk 25]
```diff
@@ -238,7 +238,23 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
 		}
 	case model.ActionDropTable, model.ActionTruncateTable:
 		tableID := job.TableID
-		startKey := tablecodec.EncodeTablePrefix(tableID)
+		// The startKey here is for compatibility with previous versions and old version did not handle endKey.
```
Comment (winkyao) [inline ddl/delete_range.go]: previous versions and old version?

File: ddl/delete_range.go [hunk 26]
```diff
@@ -238,7 +238,23 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
...
+			for _, pid := range partitionIDs {
+				startKey = tablecodec.EncodeTablePrefix(pid)
+				endKey := tablecodec.EncodeTablePrefix(pid + 1)
+				if err := doInsert(s, job.ID, pid, startKey, endKey, now); err != nil {
+					return errors.Trace(err)
+				}
+			}
+			return nil
+		}
+		startKey = tablecodec.EncodeTablePrefix(tableID)
```
Comment (winkyao) [inline ddl/delete_range.go]: if the table has partitions, we still need to add del range that encodes from tableID?

Comment (ciscoxll) [reply ddl/delete_range.go]: @winkyao No, it will return on line 255.

File: ddl/table.go [hunk 27]
```diff
@@ -213,6 +213,26 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+
+	var oldPartitionIDs []int64
+	if tblInfo.GetPartitionInfo() != nil {
+		oldPartitionIDs = getPartitionIDs(tblInfo)
+		for _, def := range tblInfo.Partition.Definitions {
+			var pid int64
+			genID := func(txn kv.Transaction) error {
```
Comment (winkyao) [inline ddl/table.go]: You can move ddl.genGlobalID to ddlCtx.genGlobalID, then you can call `ddlCtx.genGlobalID` directly.  Don't keep the same logic code in two place.

Comment (ciscoxll) [reply ddl/table.go]: @winkyao Move the code too much, do it alone with another `pr`.

File: ddl/table.go [hunk 28]
```diff
@@ -213,6 +213,26 @@ func onTruncateTable(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+
+	var oldPartitionIDs []int64
```
Comment (winkyao) [inline ddl/table.go]: Add some comments here, about why we need to gen new id.

File: ddl/db_test.go (hunk 24 above)
Comment (winkyao) [reply ddl/db_test.go]: GC run interval is `gcDefaultRunInterval = 10 * minutes`

**Pull request merged**